weasyprint/layout/__init__.py

```diff
@@ -209,6 +209,9 @@ def __init__(self, enable_hinting, style_for, get_image_from_uri,
         self.tables = {}
         self.dictionaries = {}
 
+        # TODO: this is probably the wrong place to put this
+        self.running_elements = {}
```

Comment:
we'll have to carry around the running elements somewhere. Do you think this belongs here or should I extend the state tuple?

Reply:
I think that the specification isn't clear enough. It's hard to know which running element you'll have in the page margin when multiple running elements are in the page. The `string()` function at least has a second argument for that.

Even a sigle value, in [example 5](https://drafts.csswg.org/css-gcpm/#running-syntax) for example, is not clear to me: what's on the second page?
- We could put '2 / Miranda v. Arizona in Context'.
- We could put '1 / Miranda v. Arizona in Context', as the running element is defined on page 1.
- We could put nothing, as no running element is defined on page 2.
(I think that we want the first answer.)

The place where to put running elements depends on the answer to this question. But if we think that running elements somehow behave like named strings, then it's the right place to set

Reply:
If it's not the first answer I wouldn't want to implement it because then it wouldn't be useful for me :-)
I define a [box](https://github.com/Kozea/WeasyPrint/pull/882/files#diff-bc88be45800e96e69c57218fd5c788c7R718) now that does nothing, save for [setting the running element for the page](https://github.com/Kozea/WeasyPrint/pull/882/files#diff-9f5170b54e3d607d6e89661639e531e7R671). This way, the last `running()` on the page wins, that's how I interpret the spec.

The not-doing-anything-box seems clumsy to me, can this be more elegant?

Reply:
> This way, the last `running()` on the page wins, that's how I interpret the spec.

Actually, `element()` takes an optional second argument to define which one wins (just as in `string()`).

Reply:
> This way, the last `running()` on the page wins, that's how I interpret the spec.

There's a second optional parameter for `element()` to define this. We have to use the same behavior as `string()`.

Reply:
For the record:

> I think that the specification isn't clear enough. It's hard to know which running element you'll have in the page margin when multiple running elements are in the page. The `string()` function at least has a second argument for that.

They’ve added a second parameter to the `element()` function, just as it’s done for `string()`.